### PR TITLE
Christiane Sawadogo -- Carret

### PR DESCRIPTION
**Pull request by ghost**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |   I was trying to implement the reading from the csv file but I got stuck because I had tried to implement writing to the csv file too before I realized that I couldn't make inline changes to a csv file. I found a solution (writing the changes to a new csv file and renaming it like the original) but my code was still not working so I had to revert to a more recent version to submit my assignment. I wish I had just implemented the reading from a csv file first and then experimented with the writing part.     |
| Describe a concept that you gained more clarity on as you worked on this assignment. |  As I was trying to implement the reading from a csv file requirement, I realized that if I had a database I could both read from and write to, I wouldn't need my hotel class anymore.  |
| Describe a _nominal test_ that you wrote for this assignment. |  I wrote a test to check that my reservation method would return an error if I tried to make a reservation after already reserving 20 rooms at the same date.  |
| Describe an _edge case test_ that you wrote for this assignment. | One of my tests check that the total method returns the right amount for the number of rooms and the number of days of a stay.   |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I feel like I didn't write enough pseudocode and that it would have been helpful having a more detailed plan before starting to code.  |

### CODE REVIEW

**Comment by tildeee** [comment]
Running `rake` gave me three failing tests (two from `RoomBlock` and one from `Reservation`) so be mindful and address those if possible before turning in

File: lib/hotel.rb [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    #   end
+    #   return reservations_at_date
+    # end
+
+    def available_at_date(date)
+      date = Date.parse(date.sub(/[\/]/, '-'))
+      available_rooms = []
+      @list_of_rooms.each do |room|
+        booked = false
+        unless room.booked.empty?
```
Comment (tildeee) [inline lib/hotel.rb]: this requires `Hotel` to know a lot about the internals of a `room` (that it has a `booked` hash, what the keys in that hash are, etc) and this _tightly couples_ the hotel and room to each other. This may be a place where you can transfer the responsibility of a room returning the knowledge if it's booked at a certain date into Room

File: lib/hotel.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    end
+
+    def available_at_period(arrival_date, departure_date)
+      arrival_date = Date.parse(arrival_date.sub(/[\/]/, '-'))
+      departure_date = Date.parse(departure_date.sub(/[\/]/, '-'))
+      available_rooms = []
+      @list_of_rooms.each do |room|
+        booked = false
+        room.booked.each do |stay_hash|
+          if arrival_date < stay_hash["departure"] || departure_date > stay_hash["arrival"]
```
Comment (tildeee) [inline lib/hotel.rb]: same comment as above

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      available_rooms = available_at_period(arrival_date, departure_date)
+      if find_block(client) != nil
+        raise ArgumentError.new("This client already has a room block")
+      elsif available_rooms == nil
+        raise NoAvailableRoomError.new("No avaibility at these dates.")
+      elsif available_rooms.length < number_of_rooms.to_i
+        raise NoAvailableRoomError.new("Only #{available_rooms.length} rooms available at these dates.")
+      else
+        number_of_rooms.to_i.times do |i|
+          available_at_period(arrival_date, departure_date).first.booked << {"arrival" => Date.parse(arrival_date.sub(/[\/]/, '-')), "departure" => Date.parse(departure_date.sub(/[\/]/, '-'))}
```
Comment (tildeee) [inline lib/hotel.rb]: this line is very long!

Also, instead of shoveling a very specific hash into a property of a `room`, it might be better to move this logic into the room itself. We want our room to know how to add the information to the `booked` property, not hotel

**Comment by tildeee** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| x
Answer comprehension questions	| x
**Design** |
Each class is responsible for a single piece of the program | x
Classes are loosely coupled | see comments
**Wave 1** |
List rooms | x
Reserve a room for a given date range | x
List reservations for a given date | x
Calculate reservation price | x
Invalid date range produces an error | only checks in `Reservation#stay_length` which doesn't get called anywhere
Test coverage | x
**Wave 2** |
View available rooms for a given date range | x
Reserving a room that is not available produces an error | x
Test coverage | x
**Wave 3** |
Create a block of rooms | x
Check if a block has rooms | x
Reserve a room from a block | x
Test coverage | x
**Additional Feedback** |

File: specs/hotel_spec.rb [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+      @hotel.list_of_rooms[3].booked.must_be_empty
+      @hotel.list_of_rooms[4].booked.must_be_empty
+      @hotel.list_of_rooms[5].booked.must_be_empty
+      @hotel.make_reservation("Alex luedtke", "2017/9/17", "2017/9/19", 1)
+      @hotel.list_of_rooms[0].booked[0]["arrival"].class.must_equal Date
+      @hotel.list_of_rooms[1].booked[0]["arrival"].class.must_equal Date
+      @hotel.list_of_rooms[2].booked[0]["arrival"].class.must_equal Date
+      @hotel.list_of_rooms[3].booked[0]["arrival"].class.must_equal Date
+      @hotel.list_of_rooms[4].booked.must_be_empty
+      @hotel.list_of_rooms[5].booked.must_be_empty
```
Comment (tildeee) [inline specs/hotel_spec.rb]: Most of these lines (lines 53-77) are testing for the properties on room.

In the `Hotel` test, we want to test for things that are closely related to hotel, and in the `Room` test, we want to test for things that are closely related to room. It's not bad to test for the properties of a given room in `@hotel.list_of_rooms` but it's a small code smell to need to specify several layers into `room` what its internals are.

**Comment by tildeee** [comment]
Great work!

In general, your design is pretty okay, but your hotel has to alter/know a lot about the internals of a room (like its `booked` property). Try making some methods in `Room` that hotel can call, so that Room updates itself on its own.

As a general rule, if there's a lot of _dot chaining_ going on in one line, then a class may be doing too much to another object.


Lastly, I'm so excited you got to make a CLI file for this! But when I put in the dates, there must be some sort of bug around parsing the dates, because it never works for me! Darn!

**Comment by ghost** [comment]
I started back to zero because my original project was in a confusing state. I only implemented the Room class so far. I'd like to finish implementing this project to make sure I understood the single responsibility principle and still remember pre-Rails Ruby, if it's ok. Sorry for turning it so late.
